Post: Matryoshka

marczych opened the pull request:
I figured it's about time I wrote a blog post about Matryoshka
explaining why we wrote it and how we use it.

Any feedback is appreciated!

@davidrans @danielbeardsley

marczych committed:
Post: Matryoshka

I figured it's about time I wrote a blog post about Matryoshka
explaining why we wrote it and how we use it.

Any feedback is appreciated!

davidrans commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,147 @@
+// then results in a key of '0fb4ae36-key'.
+$cache->set('key', 'value');
+$cache->set('key2', 'value2');
+$value = $cache->get('key'); // '0fb4ae36-key2' => 'value'
```
I think this comment should be:

``` php
// '0fb4ae36-key' => 'value'`.
```

That is, `key` instead of `key2`.

marczych replied:
Good catch. Fixed with 945907d78e752b4b6cdced4dcf1cc5a39349cc51.

davidrans commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,147 @@
+$cache = new Matryoshka\Scope($memcachedBackend, 'name');
+
+// This results in a get request to memcached for 'scope-name'
+// which results in something like '0fb4ae36'. This `set` call
```
This comment seems to me to either need more detail or less. What do you think about:

> This creates a Backend where each new key that is `set` is prepended with a random prefix like '0fb4ae36'.

marczych replied:
What I have now makes sense to me but it can probably be improved. I want to make it clear that the random prefix is stored in memcached which makes it tricky.

danielbeardsley replied:
I think the current text is fine.

marczych committed:
Matryoshka post: Fix scope key comment

davidrans commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,147 @@
+`getAndSet` makes populating the cache dead simple:
+
+{% highlight php startinline=true %}
+$cache = new Matryoshka\Ephemeral();
```
Since you haven't mentioned `Ephemeral` before this, it could be confusing to readers.  They may infer that the helper functions are associated with this Backend.

marczych replied:
Yeah, you're right. Just remove it I guess? Referencing `$cache` down below without defining it in this block is probably fine.

davidrans replied:
Yeah, you could use `Matryoshka\Memcache::create($memcache)`, but they probably get the idea by now.

marczych replied:
Fixed with 36e3e33b3051a5d9e53363ea15140c1893073527.

davidrans replied:
Looks like there's still one instance of `Ephemeral` here

marczych replied:
Fixed with a7212109b2efb60cd206571629bc0ea8549ac546.

davidrans commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,147 @@
+$keys = [
+   'key1' => ['id1a', 'id1b'],
+   'key2' => ['id2a', 'id2b']
+];
```
What do you think of using a simpler example here. `id2a` almost looks like a random hash at first glance.
Perhaps:

``` php
$keys = [
   'key1' => 1,
   'key2' => 2
];
```

marczych replied:
I had `id1` and `id2` before but I like having an array so it's obvious that it's purely for the callback and Matryoshka doesn't care what it is. I guess the key needs to be updated to make the example more compelling. How about:

``` php
$keys = [
   'key-1-a' => [1, 'a'],
   'key-2-b' => [2, 'b']
];
```

davidrans replied:
Yeah, I think that's better!

marczych replied:
Fixed with 37965d32c33ab3c06929b54d5baf025b7f442b25.

davidrans commented:
This is awesome! It'll be so cool if the community starts to contribute.
CR :coffee:

marczych committed:
Matryoshka: Remove Ephemeral reference

It's not necessary and might lead to confusion.

davidrans commented:
CR :coffee:

marczych committed:
Matryoshka: Update keys for multi-get example

marczych committed:
Style: Increase inline code font size

It was noticeably smaller than the rest of the body text which was
distracting.

marczych commented:
I think I've addressed all of your concerns, @davidrans. Thanks for the feedback!

I also updated the style of inline code slightly: c4023f06aa9a4a8c271c5f483768b619fa2ebe79

Before:
![before](https://cloud.githubusercontent.com/assets/1094445/5729700/35b82b8c-9b29-11e4-99a3-7ce3345aed06.png)

After:
![after](https://cloud.githubusercontent.com/assets/1094445/5729701/35bd1020-9b29-11e4-864b-ea1484871610.png)

xiongchiamiov commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,146 @@
+Similarly, `getAndSetMultiple` makes doing multi-gets significantly easier:
+
+{% highlight php startinline=true %}
+// Array of key => id. The id's can be anything used to identify
```
I think this [should be](https://en.wiktionary.org/wiki/ids#English) `ids`, without the apostrophe.

marczych replied:
Right you are! Fixed with 253b9df7138aafded4460ec163a4926ba9e54910.

xiongchiamiov commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,146 @@
+// caller. The values are returned in the same order as the
+// provided keys.
+$values = $cache->getAndSetMultiple($keys, function($missing) {
+   // Use the id's to fill in the missing values.
```
There's another one here.

xiongchiamiov commented:
I'm familiar with the project (we talked about it at length while you were designing it), but I think it's a pretty good explanation, and a well-written piece overall!

marczych committed:
Matryoshka: id's => ids

marczych committed:
Matryoshka: Remove straggling Ephemeral reference

marczych commented:
Thanks for all the great feedback, @xiongchiamiov and @davidrans!

I'd like to publish this sooner rather than later. Mind taking at least a cursory glance at this, @danielbeardsley?

danielbeardsley commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,145 @@
+Like most websites, we make heavy use of caching to reduce load on our servers and decrease page response times.
+Our caching daemon of choice is [memcached].
+The PHP extensions are certainly usable and provide all of the core functionality that you could need.
+However, we use a lot of patterns to make our day to day caching much easier that aren't provided by the extensions.
```
`s/that/which/` ?

danielbeardsley commented on _posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md:
```diff
@@ -0,0 +1,145 @@
+The PHP extensions are certainly usable and provide all of the core functionality that you could need.
+However, we use a lot of patterns to make our day to day caching much easier that aren't provided by the extensions.
+
+In comes [Matryoshka], an open source caching library for PHP which makes common operations easier and allows for on-the-fly configuration.
```
Maybe "So we developed..." instead of "In comes..."

I think we should take a bit more credit here, you wrote it from scratch.
+1 - David

danielbeardsley commented:
Even without my suggestions, this is a go. CR :ok:

MutantFreak committed:
Minor grammar and phrasing fixes

Marc says this is plenty of changes for one commit.

MutantFreak committed:
Just some day-to-day grammar fixes

Because grammar.

marczych committed:
Matryoshka: Update publish date

marczych commented:
Thanks again for all feedback! It's now [live](http://itbrokeand.ifixit.com/2015/01/20/matryoshka-configurable-caching-library-for-php.html)!
